// carbon-components/es/components/text-area/text-area.js
function _typeof(obj) {
  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function _typeof(obj) {
      return typeof obj;
    };
  } else {
    _typeof = function _typeof(obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    };
  }

  return _typeof(obj);
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _possibleConstructorReturn(self, call) {
  if (call && (_typeof(call) === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}
/**
 * Copyright IBM Corp. 2016, 2018
 *
 * This source code is licensed under the Apache-2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */


import settings from '../../globals/js/settings';
import mixin from '../../globals/js/misc/mixin';
import createComponent from '../../globals/js/mixins/create-component';
import initComponentBySearch from '../../globals/js/mixins/init-component-by-search';
import handles from '../../globals/js/mixins/handles';
import on from '../../globals/js/misc/on';

var TextArea =
/*#__PURE__*/
function (_mixin) {
  _inherits(TextArea, _mixin);
  /**
   * Text Area.
   * @extends CreateComponent
   * @extends InitComponentBySearch
   * @extends Handles
   * @param {HTMLElement} element - The textarea element
   */


  function TextArea(_element, options) {
    var _this;

    _classCallCheck(this, TextArea);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(TextArea).call(this, _element, options));

    _this._handleInput = function (_ref) {
      var element = _ref.element,
          length = _ref.length;
      element.querySelector(_this.options.selectorCharCounter).textContent = length;
    };

    _this.manage(on(_this.element, 'input', function (event) {
      if (event.target.maxLength < 0) {
        return;
      }

      _this._handleInput({
        element: _element,
        length: event.target.value.length
      });
    }));

    return _this;
  }
  /**
   * Updates the character counter
   * @param {Object} obj - The elements that can change in the component
   * @param {HTMLElement} obj.element - The textarea element
   * @param {HTMLElement} obj.length - The length of the textarea value
   */


  _createClass(TextArea, null, [{
    key: "options",

    /**
     * The component options.
     *
     * If `options` is specified in the constructor,
     * {@linkcode TextArea.create .create()},
     * or {@linkcode TextArea.init .init()},
     * properties in this object are overriden for the instance being
     * created and how {@linkcode TextArea.init .init()} works.
     * @property {string} selectorInit The CSS selector to find textarea UIs.
     */
    get: function get() {
      var prefix = settings.prefix;
      return {
        selectorInit: '[data-text-area]',
        selectorCharCounter: ".".concat(prefix, "--text-area--character-counter--length")
      };
    }
    /**
     * The map associating DOM element and textarea UI instance.
     * @type {WeakMap}
     */

  }]);

  TextArea.components = new WeakMap();
  return TextArea;
}(mixin(createComponent, initComponentBySearch, handles));

export { TextArea as default };
